kafka.go

```diff
@@ -1,161 +1,215 @@
+		logrus.Errorf("unable to construct event : %s , error : %v", publishBuilder.eventName, err)
+		return
 	}
-}
 
-type eventEntry struct {
-	Event
+	for _, topic := range publishBuilder.topic {
+		config := client.publishConfig
+		config.Topic = topic
+		writer := kafka.NewWriter(config)
 
-	encoded []byte
-	err     error
+		err = writer.WriteMessages(publishBuilder.ctx, message)
+		if err != nil {
+			logrus.Errorf("unable to publish event to kafka. topic : %s , error %v : ", topic, err)
+		}
+	}
 }
 
-func (evtEntry *eventEntry) ensureEncoded() {
-	if evtEntry.encoded == nil && evtEntry.err == nil {
-		evtEntry.encoded, evtEntry.err = json.Marshal(evtEntry)
+// constructEvent construct event message
+func constructEvent(publishBuilder *PublishBuilder) (kafka.Message, error) {
+	id := generateID()
+	event := &Event{
+		ID:        id,
+		EventName: publishBuilder.eventName,
+		Namespace: publishBuilder.namespace,
+		ClientID:  publishBuilder.clientID,
+		UserID:    publishBuilder.userID,
+		Timestamp: time.Now().UTC(),
+		Version:   version,
+		Payload:   publishBuilder.payload,
 	}
-}
 
-// Length returns the length of the event entry
-func (evtEntry *eventEntry) Length() int {
-	evtEntry.ensureEncoded()
-	return len(evtEntry.encoded)
+	eventBytes, err := marshal(event)
+	if err != nil {
+		return kafka.Message{}, err
+	}
+
+	return kafka.Message{
+		Key:   []byte(id),
+		Value: eventBytes,
+	}, nil
+
 }
 
-// Encode serialize the struct
-func (evtEntry *eventEntry) Encode() ([]byte, error) {
-	evtEntry.ensureEncoded()
-	return evtEntry.encoded, evtEntry.err
+// Register register callback and consume a topic
+func (client *KafkaClient) Register(subscribeBuilder *SubscribeBuilder) {
+	go func() {
+		topic := constructTopic(client.realm, subscribeBuilder.topic)
+		if client.registerCallback(topic, subscribeBuilder.eventName, subscribeBuilder.callback) {
+			logrus.Warnf("topic and event already registered. event : %s", subscribeBuilder.eventName)
+			return
+		}
+
+		config := client.subscribeConfig
+		config.Topic = subscribeBuilder.topic
+		reader := kafka.NewReader(config)
+		if err := reader.SetOffset(kafka.LastOffset); err != nil {
+			logrus.Error("unable to set offset : ", err)
+			return
+		}
+
+		defer func() {
+			_ = reader.Close()
+		}()
+		for {
+			consumerMessage, err := reader.ReadMessage(subscribeBuilder.ctx)
+			if err != nil {
+				logrus.Error("unable to consume topic from kafka : ", err)
+				return
+			}
+			go client.processMessage(consumerMessage)
+		}
+	}()
 }
 
-// PublishEvent push an event to a Kafka topic synchronously
-func (client *KafkaClient) PublishEvent(event *Event) error {
-	event.Realm = client.realm
-	event.Time = time.Now().UTC()
+// registerCallback add callback to map with topic and eventName as a key
+func (client *KafkaClient) registerCallback(topic, eventName string, callback func(event *Event, err error)) (
+	isRegistered bool) {
+	client.RWMutex.Lock()
+	defer client.RWMutex.Unlock()
```

Comment:
do you use mutex because of the map? can the map be replaced with `sync.Map` instead?

Reply:
based on this [article](https://medium.com/@deckarep/the-new-kid-in-town-gos-sync-map-de24a6bf7c2c), 
rwmutex have a better performance than sync map in a single core processing and
sync map would be great if we use multi-thread / multiple worker and for this case, I think we dont need to make it multi-thread. what do you think?

Reply:
if this is not multi-threaded, why are we using mutex?

Reply:
Ah I see, I misunderstand about it and this case would be working on multiple goroutines which means map.sync it would be better. thanks

Reply:
from the article you mentioned tho, sync.map only faster than mutex (especially if you separate lock and rlock) if the number of goroutine is high. so "better" is depending on the number of goroutines you have trying to access at the same time

Reply:
yes, you're right. 
To find out the number of goroutine .
I tried to create a benchmark that reflect with our case using 10000x read and 100 write operation and the result is sync.Map have a ~1x faster than RWmutex but needs memory allocation ~10x. 
but If we increase the read until 100000x and 100 write, sync.Map still faster and take less memory than RWmutex.  
With this result, I prefer to use sync.Map because the differences are not significant

Reply:
awesome